docs/datatypes.md

```diff
@@ -0,0 +1,253 @@
+
+Defining New Types
+==================
+
+Unlike most Lisps, Irken's type system is based on 'algebraic
+datatypes', where complex types are built from simpler types.
+
+New types are created with the `datatype` special form:
+
+    (datatype year
+      (:known int)
+      (:approx int)
+      (:range int int)
+      (:unknown)
+      )
+
+Here, we've created a new datatype `year`, which has four
+different 'variants'.  To create a 'known' year:
+
+    (year:known 1972)
+
+[`year:known` is called a 'constructor'.  It builds a value of
+of datatype `year`, with the variant `known`.]
+
+To create a 'range' year...
+
+    (year:range 1530 1540)
+
+... and an 'unknown' year:
+
+    (year:unknown)
+
+Another example:
+
+    (datatype monarch
+      (:king bool)
+      (:pharaoh bool)
+      (:machine)
+      )
+
+Here, we use a `bool` to indicate male or female, where:
+
+    (monarch:king #f)
+
+would indicate a queen.
+
+We can combine the two dataypes:
+
+    (datatype rule
+      (:reign monarch year year)
+      (:chaos year year)
+      )
+
+The rule of our robot overlords:
+
+    (rule:reign (monarch:machine) (year:approx 2050) (year:unknown))
+
+Polymorphic Datatypes
+---------------------
+
+We need not use a specific type when defining a new datatype.
+Instead, we can define it to work with any type:
+
+    (datatype thing
+      (:one 'a)
+      (:two 'a 'a)
+      )
+
+Here, `'a` indicates 'any type'.  We can now create new `things` of
+different types:
+
+    (thing:one "a string")
+    (thing:one #f)
+    (thing:two 27 27)
+
+When we create such a value, we 'instantiate' the polymorphic type
+(i.e. 'fill in the blanks').  We say that `(thing:one "a string")` is
+of type `(thing string)` and `(thing:two 27 27)` is of type `(thing int)`.
+
+Note that we _cannot_ do this:
+
+    (thing:two 10 #f)
+
+Because `10` and `#f` are different types (`int` and `bool`,
+respectively), and there's no way to 'fill in' two different types for
+`'a`.
+
+If we need a datatype that allowed this, we might use:
+
+    (datatype thing2
+      (:one 'a)
+      (:two 'a 'b)
+      )
+
+We say that `thing2` has type `(thing2 'a 'b)`.
+
+Note: there is still a constraint on `thing2` objects.  These two
+are _not_ of the same type:
+
+    (thing2:one 34)
+    (thing2:two #f "string")
+
+But these two _are_ of the same type:
+
+    (thing2:one 34)
+    (thing2:two 19 "string")
```

Comment:
Why? because the _first_ value of the second instance is an int? because it contains an int?